File: VodArchiver/VodArchiver/VideoInfo/IVideoInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace VodArchiver.VideoInfo {
	public enum StreamService {
		Unknown,
		Twitch,
		Hitbox,
		TwitchChatReplay,
		Youtube,
		RawUrl,
		FFMpegJob,
	}

	public enum RecordingState {
		Unknown,
		Live,
		Recorded,
	}

	public enum VideoFileType {
		M3U,
		FLVs,
		Unknown
	}

	public abstract class IVideoInfo {
		public virtual StreamService Service { get; set; }
		public virtual string Username { get; set; }
		public virtual string VideoId { get; set; }
		public virtual string VideoTitle { get; set; }
		public virtual string VideoGame { get; set; }
		public virtual DateTime VideoTimestamp { get; set; }
		public virtual TimeSpan VideoLength { get; set; }
		public virtual RecordingState VideoRecordingState { get; set; }
		public virtual VideoFileType VideoType { get; set; }

		public abstract XmlNode Serialize( XmlDocument document, XmlNode node );

		public static IVideoInfo Deserialize( XmlNode node ) {
			string type = node.Attributes["_type"].Value;
			switch ( type ) {
				case "GenericVideoInfo": return new GenericVideoInfo( node );
				case "HitboxVideoInfo": return new HitboxVideoInfo( node );
				case "TwitchVideoInfo": return new TwitchTwixelVideoInfo( node );
				case "YoutubeVideoInfo": return new YoutubeVideoInfo( node );
				case "FFMpegReencodeJobVideoInfo": return new FFMpegReencodeJobVideoInfo( node );
				default: throw new Exception( "Unknown video job type: " + type );
			}
		}

		public override bool Equals( object obj ) {
			return Equals( obj as IVideoInfo );
		}

		public bool Equals( IVideoInfo other ) {
			return other != null && Service == other.Service && VideoId == other.VideoId;
		}

		public override int GetHashCode() {
			return Service.GetHashCode() ^ VideoId.GetHashCode();
		}
	}
}

File: native/vodarchiver/videoinfo/i-video-info.cpp
#include "i-video-info.h"

namespace VodArchiver {
std::string_view StreamServiceToString(StreamService service) {
    switch (service) {
        case StreamService::Twitch: return "Twitch";
        case StreamService::Hitbox: return "Hitbox";
        case StreamService::TwitchChatReplay: return "TwitchChatReplay";
        case StreamService::Youtube: return "Youtube";
        case StreamService::RawUrl: return "RawUrl";
        case StreamService::FFMpegJob: return "FFMpegJob";
        default: return "Unknown";
    }
}

std::optional<StreamService> StreamServiceFromString(std::string_view sv) {
    if (sv == "Unknown") {
        return StreamService::Unknown;
    } else if (sv == "Twitch") {
        return StreamService::Twitch;
    } else if (sv == "Hitbox") {
        return StreamService::Hitbox;
    } else if (sv == "TwitchChatReplay") {
        return StreamService::TwitchChatReplay;
    } else if (sv == "Youtube") {
        return StreamService::Youtube;
    } else if (sv == "RawUrl") {
        return StreamService::RawUrl;
    } else if (sv == "FFMpegJob") {
        return StreamService::FFMpegJob;
    }
    return std::nullopt;
}

std::string_view RecordingStateToString(RecordingState state) {
    switch (state) {
        case RecordingState::Live: return "Live";
        case RecordingState::Recorded: return "Recorded";
        default: return "Unknown";
    }
}

std::optional<RecordingState> RecordingStateFromString(std::string_view sv) {
    if (sv == "Unknown") {
        return RecordingState::Unknown;
    } else if (sv == "Live") {
        return RecordingState::Live;
    } else if (sv == "Recorded") {
        return RecordingState::Recorded;
    }
    return std::nullopt;
}

std::string_view VideoFileTypeToString(VideoFileType type) {
    switch (type) {
        case VideoFileType::M3U: return "M3U";
        case VideoFileType::FLVs: return "FLVs";
        default: return "Unknown";
    }
}

std::optional<VideoFileType> VideoFileTypeFromString(std::string_view sv) {
    if (sv == "M3U") {
        return VideoFileType::M3U;
    } else if (sv == "FLVs") {
        return VideoFileType::FLVs;
    } else if (sv == "Unknown") {
        return VideoFileType::Unknown;
    }
    return std::nullopt;
}

IVideoInfo::~IVideoInfo() = default;
} // namespace VodArchiver
